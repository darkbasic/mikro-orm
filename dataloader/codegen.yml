overwrite: true
schema: './src/schema/typeDefs/*.graphql'
documents: null
generates:
  ./types/graphql.d.ts:
    plugins:
      - 'typescript'
      - 'typescript-resolvers'
    config:
      contextType: ../src#MyContext
      mappers:
        User: ../src/entities/User#UserEntity
        Site: ../src/entities/Site#SiteEntity
        Sport: ../src/entities/Sport#SportEntity
        Specialty: ../src/entities/Specialty#SpecialtyEntity
        Match: ../src/entities/Match#MatchEntity
      scalars:
        Date: Date
        URL: string
        EmailAddress: string
  ./graphql.schema.json:
    plugins:
      - 'introspection'
