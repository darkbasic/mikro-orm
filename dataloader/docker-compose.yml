version: '3.4'

services:
  postgre:
    container_name: postgre12
    image: postgres:12.2
    ports:
      - 5432:5432
    volumes:
      - postgre12:/var/lib/postgresql/data
    environment:
      POSTGRES_HOST_AUTH_METHOD: trust

volumes:
  postgre12:
