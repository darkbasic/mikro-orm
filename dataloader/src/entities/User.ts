import {
  Entity,
  Property,
  Enum,
  OneToMany,
  Collection,
  ManyToMany,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Match} from './Match';

export enum Sex {
  MALE,
  FEMALE,
}

interface UserContructor {
  id?: number;
  email: string;
  name: string;
  surname: string;
  sex: Sex;
  password: string;
  picture?: string;
}

export {User as UserEntity};

@Entity()
export class User extends BaseEntity {
  @Property()
  email: string;

  @Property()
  name: string;

  @Property()
  surname: string;

  @Enum()
  sex: Sex;

  @Property()
  password: string;

  @Property()
  picture?: string;

  @OneToMany(() => Match, match => match.creator)
  createdMatches: Collection<Match> = new Collection<Match>(this);

  @ManyToMany(() => Match, match => match.partecipants)
  partecipatedMatches: Collection<Match> = new Collection<Match>(this);

  constructor({
    id,
    email,
    name,
    surname,
    sex,
    password,
    picture,
  }: UserContructor) {
    super({id});
    this.email = email;
    this.name = name;
    this.surname = surname;
    this.sex = sex;
    this.password = password;
    if (picture) {
      this.picture = picture;
    }
  }
}
