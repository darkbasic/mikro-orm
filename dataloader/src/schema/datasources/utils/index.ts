import {GraphQLResolveInfo} from 'graphql';
import {
  Reference,
  AnyEntity,
  EntityManager,
  Primary,
  IdentifiedReference,
  Collection,
  WrappedEntity,
  wrap,
  EntityRepository,
} from 'mikro-orm';
import DataLoader from 'dataloader';

export function ensureIsNum(id: string | number): number;
export function ensureIsNum(
  id: string | number | null | undefined
): number | undefined;
export function ensureIsNum(ids: Array<string | number>): number[];
export function ensureIsNum(
  ids: Array<string | number> | null | undefined
): number[] | undefined;
export function ensureIsNum(
  idOrIds: null | undefined | string | number | (string | number)[]
): undefined | number | number[] {
  if (idOrIds === null || idOrIds === undefined) {
    return undefined;
  }
  if (idOrIds instanceof Array) {
    return idOrIds.map(id => {
      const num = Number(id);
      if (isNaN(num)) {
        throw new Error(`The ID ${id} is not a valid number`);
      }
      return num;
    });
  } else {
    const num = Number(idOrIds);
    if (isNaN(num)) {
      throw new Error(`The ID ${idOrIds} is not a valid number`);
    }
    return num;
  }
}

export function shouldPopulate({fieldNodes}: GraphQLResolveInfo): boolean {
  if (fieldNodes.length === 1) {
    const selections = fieldNodes[0].selectionSet?.selections;
    if (selections?.length === 1) {
      if (selections[0].kind !== 'InlineFragment') {
        return selections[0].name.value !== 'id';
      }
    }
  }
  return true;
}

export function getItemNoPopulate<T>(
  ref: Reference<T>,
  info?: GraphQLResolveInfo
): T | undefined {
  return !info || shouldPopulate(info) ? undefined : ((ref as unknown) as T);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getAllPropertyNames(obj: any) {
  const props: string[] = [];

  do {
    Object.getOwnPropertyNames(obj).forEach(prop => {
      if (props.indexOf(prop) === -1) {
        props.push(prop);
      }
    });
  } while ((obj = Object.getPrototypeOf(obj)));

  return props;
}

interface Id {
  id: number | string;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasRef<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity>>;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasCol<T>(
  entity: T
): T & WrappedEntity<T, keyof T> & Record<string, Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, Collection<AnyEntity>>;
}

function hasRefOrCol<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>>;
}

function isRef<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is IdentifiedReference<T> {
  return !(refOrCol instanceof Collection);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isCol<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is Collection<T> {
  return refOrCol instanceof Collection;
}

type RepoFind<K> = {
  repo: EntityRepository<K>;
  filter: Record<string, Primary<K> | Primary<K>[]>;
  many: boolean;
};

type Result<K> = {
  entityName: string;
  keys: string[];
  entitiesOrError: K[] | Error;
};

function toArray<T>(arrOrAny: T | T[]): T[] {
  return Array.isArray(arrOrAny) ? arrOrAny : [arrOrAny];
}

function objectsHaveSameKeys<T>(
  ...objects: Record<string, Primary<T> | Primary<T>[]>[]
): boolean {
  const allKeys = new Set(
    objects.reduce(
      (keys, object) => keys.concat(Object.keys(object)),
      [] as string[]
    )
  );
  return objects.every(object => allKeys.size === Object.keys(object).length);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayAreEqual<T extends Primary<any>[] | string[]>(
  a: T,
  b: T
): boolean {
  return JSON.stringify(a.sort()) === JSON.stringify(b.sort());
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayIncludes<T extends (string | Primary<any>)[]>(
  arr: T,
  target: T
): boolean {
  return target.every(v => arr.includes(v));
}

export function groupPrimaryKeysByEntity<T extends AnyEntity<T>>(
  refs: readonly Reference<T>[]
) {
  return refs.reduce((acc, cur) => {
    const {className} = cur.__meta;
    const primaryKeys: Primary<T>[] = acc[className];
    const primaryKey: Primary<T> = cur.__primaryKey;
    acc[className] = primaryKeys
      ? [...new Set([...primaryKeys, primaryKey])]
      : [primaryKey];
    return acc;
  }, {} as Record<string, Primary<T>[]>);
}

export function groupInversedOrMappedKeysByEntity<T extends AnyEntity<T>>(
  collections: readonly Collection<T>[]
): Record<string, Record<string, Primary<T>[]>> {
  return collections.reduce((acc, cur) => {
    const className = cur.property.type;
    const record = acc[className] || (acc[className] = {});
    // Many to Many vs One to Many
    const inversedOrMappedBy = cur.property.inversedBy
      ? 'inversedBy'
      : cur.property.mappedBy
      ? 'mappedBy'
      : undefined;
    if (!inversedOrMappedBy) {
      throw new Error('Cannot find inversedBy or mappedBy prop');
    }
    const key: Primary<T> = cur.owner.__primaryKey;
    const keys: Primary<T>[] = record[cur.property[inversedOrMappedBy]];
    record[cur.property[inversedOrMappedBy]] = keys
      ? [...new Set([...keys, key])]
      : [key];
    return acc;
  }, {} as Record<string, Record<string, Primary<T>[]>>);
}

export function groupFindQueries<T extends AnyEntity<T>>(
  repoFinds: readonly RepoFind<T>[]
) {
  return repoFinds.reduce((acc, {repo, filter}) => {
    const entityName = repo['entityName'] as string;
    if (acc[entityName]) {
      let matchFound = false;
      const curFilters = acc[entityName].map(el => ({...el}));
      for (const curFilter of curFilters) {
        if (objectsHaveSameKeys(curFilter, filter)) {
          matchFound = true;
          Object.entries(filter).forEach(([prop, idOrIds]) => {
            const curIdOrIds = curFilter[prop];
            if (
              Array.isArray(curIdOrIds) ||
              Array.isArray(idOrIds) ||
              (curIdOrIds && curIdOrIds !== idOrIds)
            ) {
              curFilter[prop] = [
                ...new Set([...toArray(curIdOrIds ?? []), ...toArray(idOrIds)]),
              ];
            } else {
              curFilter[prop] = idOrIds;
            }
          });
          break;
        }
      }
      if (matchFound) {
        acc[entityName] = [...curFilters];
      } else {
        acc[entityName] = [...curFilters, filter];
      }
    } else {
      acc[entityName] = [filter];
    }
    return acc;
  }, {} as Record<string, Record<string, Primary<T> | Primary<T>[]>[]>);
}

export class EntityDataLoader<T extends Id = Id, K = AnyEntity<T, 'id'>> {
  private bypass: boolean;
  private refLoader: DataLoader<IdentifiedReference<K>, K>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private colLoader: DataLoader<Collection<any>, K[]>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private findLoader: DataLoader<RepoFind<any>, K[] | K | null>;

  constructor(em: EntityManager, bypass = false) {
    this.bypass = bypass;

    this.refLoader = new DataLoader<IdentifiedReference<K>, K>(async refs => {
      const groupedIds = groupPrimaryKeysByEntity(refs);
      const promises = Object.entries(groupedIds).map(([entity, ids]) =>
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        em.getRepository<K>(entity).find(ids as any)
      );
      await Promise.all(promises);
      return Promise.all(refs.map(async ref => await ref.load()));
    });

    this.colLoader = new DataLoader<Collection<K>, K[]>(async collections => {
      const groupedKeys = groupInversedOrMappedKeysByEntity(collections);
      const promises: Promise<K[]>[] = Object.entries(
        groupedKeys
      ).map(([entity, record]) =>
        em.getRepository<K>(entity).find(record, Object.keys(record))
      );
      const results = (await Promise.all(promises))
        .flat()
        .map(entity => wrap(entity));
      return collections.map(collection =>
        results.filter(result => {
          // Class matches
          if (result.__meta.className === collection.property.type) {
            const inversedOrMappedBy = collection.property.inversedBy
              ? 'inversedBy'
              : collection.property.mappedBy
              ? 'mappedBy'
              : undefined;
            if (!inversedOrMappedBy) {
              throw new Error('Cannot find inversedBy or mappedBy prop');
            }
            const refOrCol = hasRefOrCol(result)[
              collection.property[inversedOrMappedBy]
            ];
            return isRef(refOrCol)
              ? refOrCol.__primaryKey === collection.owner.__primaryKey
              : refOrCol
                  .getItems()
                  .map(el => el.__primaryKey)
                  .includes(collection.owner.__primaryKey);
          }
          return false;
        })
      );
    });

    this.findLoader = new DataLoader<RepoFind<K>, K[] | K | null>(
      async repoFinds => {
        const groupedKeys = groupFindQueries(repoFinds);
        const groupedResults: Result<K>[] = await Promise.all(
          Object.entries(groupedKeys)
            .map(([entityName, records]) => {
              return records.map(async record => {
                let entitiesOrError: K[] | Error;
                try {
                  entitiesOrError = await em
                    .getRepository<K>(entityName)
                    .find(record, Object.keys(record));
                } catch (e) {
                  entitiesOrError = e as Error;
                }
                return {
                  entityName,
                  keys: Object.keys(record),
                  entitiesOrError,
                };
              });
            })
            .flat()
        );

        return repoFinds.map(({repo, filter, many}) => {
          const result = groupedResults.find(
            ({entityName, keys, entitiesOrError}) => {
              if (
                entityName === String(repo['entityName']) &&
                arrayAreEqual(Object.keys(filter), keys)
              ) {
                if (!(entitiesOrError instanceof Error)) {
                  return true;
                } else {
                  throw entitiesOrError;
                }
              }
              return false;
            }
          );
          if (!result) {
            throw new Error('Cannot match results');
          }

          const {entitiesOrError} = result;
          if (!(entitiesOrError instanceof Error)) {
            return (
              entitiesOrError[many ? 'filter' : 'find'](entity => {
                for (const key of Object.keys(filter)) {
                  if (
                    !arrayIncludes(
                      toArray(filter[key]),
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      toArray((entity as any)[key]).map(
                        el => el.id as Primary<K>
                      )
                    )
                  ) {
                    return false;
                  }
                }
                return true;
              }) ?? null
            );
          } else {
            throw entitiesOrError;
          }
        });
      }
    );
  }

  load<L extends K>(ref: IdentifiedReference<L>, bypass?: boolean): Promise<L>;
  load<L extends K>(collection: Collection<L>, bypass?: boolean): Promise<L[]>;
  load<L extends K>(
    refOrCol: IdentifiedReference<L> | Collection<L>,
    bypass?: boolean
  ): Promise<L> | Promise<L[]> {
    if (isRef(refOrCol)) {
      return bypass ?? this.bypass
        ? refOrCol.load()
        : (this.refLoader.load(refOrCol) as Promise<L>);
    } else {
      return bypass ?? this.bypass
        ? refOrCol.loadItems()
        : (this.colLoader.load(refOrCol) as Promise<L[]>);
    }
  }

  find<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L[]> {
    return bypass ?? this.bypass
      ? repo.find(filter)
      : (this.findLoader.load({repo, filter, many: true}) as Promise<L[]>);
  }

  findOne<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L | null> {
    return bypass ?? this.bypass
      ? repo.findOne(filter)
      : (this.findLoader.load({
          repo,
          filter,
          many: false,
        }) as Promise<L | null>);
  }
}
