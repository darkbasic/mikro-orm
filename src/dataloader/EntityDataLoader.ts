/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  AnyEntity,
  IdentifiedReference,
  Collection,
  EntityManager,
  wrap,
  EntityRepository,
  Primary,
  FilterQuery,
} from '@mikro-orm/core';
import DataLoader from 'dataloader';
import {
  groupPrimaryKeysByEntity,
  groupInversedOrMappedKeysByEntity,
  groupFindQueries,
  ensureIsArray,
  RepoFind,
  Result,
  arrayAreEqual,
  arrayIncludes,
  hasRefOrCol,
  isRef,
  getRefPk,
  getHelperPk,
  getEntityMetadataClassName,
  IWrappedEntityInternalWithEntity,
} from './dataloader';

export class EntityDataLoader<T extends AnyEntity<T> = any> {
  private bypass: boolean;
  private refLoader: DataLoader<IdentifiedReference<T>, T>;
  private colLoader: DataLoader<Collection<T, AnyEntity<T>>, T[]>;
  private findLoader: DataLoader<RepoFind<T>, T[] | T | null>;

  constructor(em: EntityManager, bypass = false) {
    this.bypass = bypass;

    this.refLoader = new DataLoader<IdentifiedReference<T>, T>(async refs => {
      const groupedIds = groupPrimaryKeysByEntity(refs);
      const promises = Object.entries(groupedIds).map(([entity, ids]) =>
        (em.getRepository<T>(entity) as EntityRepository<T>).find(
          ids as FilterQuery<T>
        )
      );
      await Promise.all(promises);
      return Promise.all(refs.map(async ref => await ref.load()));
    });

    this.colLoader = new DataLoader<Collection<T, AnyEntity>, T[]>(
      async collections => {
        const groupedKeys = groupInversedOrMappedKeysByEntity(collections);
        const promises: Promise<T[]>[] = Object.entries(groupedKeys).map(
          ([entity, record]) =>
            (em.getRepository<T>(entity) as EntityRepository<T>).find(
              record,
              Object.keys(record)
            )
        );
        const results = (await Promise.all(promises))
          .flat()
          .map(
            entity => wrap(entity, true) as IWrappedEntityInternalWithEntity<T>
          );
        return collections.map(collection =>
          results
            .filter(result => {
              // Class matches
              if (
                getEntityMetadataClassName(result.__meta) ===
                collection.property.type
              ) {
                const inversedOrMappedBy = collection.property.inversedBy
                  ? 'inversedBy'
                  : collection.property.mappedBy
                  ? 'mappedBy'
                  : undefined;
                if (!inversedOrMappedBy) {
                  throw new Error('Cannot find inversedBy or mappedBy prop');
                }
                const refOrCol = hasRefOrCol(result.entity)[
                  collection.property[inversedOrMappedBy]
                ];
                return isRef(refOrCol)
                  ? getRefPk(refOrCol) ===
                      getHelperPk(collection.owner.__helper)
                  : refOrCol
                      .getItems()
                      .map(el => getHelperPk(el.__helper))
                      .includes(getHelperPk(collection.owner.__helper));
              }
              return false;
            })
            .map(wrapped => wrapped.entity)
        );
      }
    );

    this.findLoader = new DataLoader<RepoFind<T>, T[] | T | null>(
      async repoFinds => {
        const groupedKeys = groupFindQueries(repoFinds);
        const groupedResults: Result<T>[] = await Promise.all(
          Object.entries(groupedKeys)
            .map(([entityName, records]) => {
              return records.map(async record => {
                let entitiesOrError: T[] | Error;
                try {
                  const {properties} = em.getMetadata().get(entityName);
                  entitiesOrError = await (
                    em.getRepository<T>(entityName) as EntityRepository<T>
                  ).find(
                    record,
                    Object.keys(record).filter(
                      key => properties[key].reference !== 'scalar'
                    )
                  );
                } catch (e) {
                  entitiesOrError = e as Error;
                }
                return {
                  entityName,
                  keys: Object.keys(record),
                  entitiesOrError,
                };
              });
            })
            .flat()
        );

        return repoFinds.map(({repo, filter, many}) => {
          const result = groupedResults.find(
            ({entityName, keys, entitiesOrError}) => {
              if (
                entityName === String(repo['entityName']) &&
                arrayAreEqual(Object.keys(filter), keys)
              ) {
                if (!(entitiesOrError instanceof Error)) {
                  return true;
                } else {
                  throw entitiesOrError;
                }
              }
              return false;
            }
          );
          if (!result) {
            throw new Error('Cannot match results');
          }

          const {entitiesOrError} = result;
          if (!(entitiesOrError instanceof Error)) {
            return (
              entitiesOrError[many ? 'filter' : 'find'](entity => {
                for (const key of Object.keys(filter) as (keyof T & string)[]) {
                  const {properties} = em.getMetadata().get(result.entityName);
                  const idsFromSearchFilter = ensureIsArray(filter[key]);
                  const idsFromFindResult =
                    properties[key].reference !== 'scalar'
                      ? ensureIsArray(entity[key]).map(
                          (el: AnyEntity) => el.id as Primary<T>
                        )
                      : (ensureIsArray(entity[key]) as Primary<T>[]);
                  if (
                    !arrayIncludes(
                      idsFromSearchFilter,
                      idsFromFindResult,
                      entity[key] instanceof Collection ? 'any' : 'every'
                    )
                  ) {
                    return false;
                  }
                }
                return true;
              }) ?? null
            );
          } else {
            throw entitiesOrError;
          }
        });
      }
    );
  }

  load<K>(ref: IdentifiedReference<K>, bypass?: boolean): Promise<K>;
  load<K>(collection: Collection<K>, bypass?: boolean): Promise<K[]>;
  load<K>(
    refOrCol: IdentifiedReference<K> | Collection<K>,
    bypass?: boolean
  ): Promise<K> | Promise<K[]> {
    if (isRef(refOrCol)) {
      return bypass ?? this.bypass
        ? refOrCol.load()
        : (this.refLoader.load(
            refOrCol as unknown as IdentifiedReference<T>
          ) as unknown as Promise<K>);
    } else {
      return bypass ?? this.bypass
        ? refOrCol.loadItems()
        : (this.colLoader.load(
            refOrCol as unknown as Collection<T, AnyEntity<T>>
          ) as unknown as Promise<K[]>);
    }
  }

  find<K>(
    repo: EntityRepository<K>,
    filter: Record<string, number | number[]>,
    bypass?: boolean
  ): Promise<K[]> {
    return bypass ?? this.bypass
      ? repo.find(filter)
      : (this.findLoader.load({
          repo: repo as unknown as EntityRepository<T>,
          filter: filter as Record<string, Primary<T> | Primary<T>[]>,
          many: true,
        }) as unknown as Promise<K[]>);
  }

  findOne<K>(
    repo: EntityRepository<K>,
    filter: Record<string, number | number[]>,
    bypass?: boolean
  ): Promise<K | null> {
    return bypass ?? this.bypass
      ? repo.findOne(filter)
      : (this.findLoader.load({
          repo: repo as unknown as EntityRepository<T>,
          filter: filter as Record<string, Primary<T> | Primary<T>[]>,
          many: false,
        }) as unknown as Promise<K | null>);
  }

  async findOneOrFail<K>(
    repo: EntityRepository<K>,
    filter: Record<string, number | number[]>,
    bypass?: boolean
  ): Promise<K> {
    if (bypass ?? this.bypass) {
      return repo.findOneOrFail(filter);
    }
    const one = (await this.findLoader.load({
      repo: repo as unknown as EntityRepository<T>,
      filter: filter as Record<string, Primary<T> | Primary<T>[]>,
      many: false,
    })) as unknown as K | null;
    if (one == null) {
      throw new Error('Cannot find result');
    }
    return one;
  }
}
