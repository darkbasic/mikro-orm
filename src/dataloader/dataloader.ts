import {EntityMetadata} from '@mikro-orm/core';
import {
  AnyEntity,
  Reference,
  Primary,
  Collection,
  EntityRepository,
  PrimaryProperty,
} from '@mikro-orm/core';
import {IWrappedEntityInternal} from '@mikro-orm/core/dist/typings';

function objectsHaveSameKeys<T>(
  ...objects: Record<string, Primary<T> | Primary<T>[]>[]
): boolean {
  const allKeys = new Set(
    objects.reduce(
      (keys, object) => keys.concat(Object.keys(object)),
      [] as string[]
    )
  );
  return objects.every(object => allKeys.size === Object.keys(object).length);
}

export function ensureIsArray<T>(
  colOrArrOrAny: T | T[] | Collection<T, AnyEntity<T>>
): T[] {
  return colOrArrOrAny instanceof Collection
    ? colOrArrOrAny.getItems()
    : Array.isArray(colOrArrOrAny)
    ? colOrArrOrAny
    : [colOrArrOrAny];
}

export type RepoFind<K> = {
  repo: EntityRepository<K>;
  filter: Record<string, Primary<K> | Primary<K>[]>;
  many: boolean;
};

export function getEntityMetadataClassName<T extends AnyEntity<T>>(
  meta?: EntityMetadata<T>
): string {
  if (!meta) {
    throw new Error('Cannot find entity metadata');
  }
  return meta.className;
}

export function getRefClassName<T extends AnyEntity<T>>(
  ref: Reference<T>
): string {
  return getEntityMetadataClassName((ref['entity'] as T).__meta);
}

export function getHelperPk<T extends AnyEntity<T>>(
  helper?: IWrappedEntityInternal<T>
): Primary<T> {
  if (!helper) {
    throw new Error('Cannot find entity helper');
  }
  if (helper.__primaryKeys.length !== 1) {
    throw new Error('Composite PKs not yet supported');
  }
  return helper.__primaryKeys[0];
}

export function getRefPk<T extends AnyEntity<T>>(
  ref: Reference<T>
): Primary<T> {
  return getHelperPk((ref['entity'] as T).__helper);
}

export function groupPrimaryKeysByEntity<T extends AnyEntity<T>>(
  refs: readonly Reference<T>[]
) {
  return refs.reduce((acc, cur) => {
    const className = getRefClassName(cur);
    const primaryKeys: Primary<T>[] = acc[className];
    const primaryKey: Primary<T> = getRefPk(cur);
    acc[className] = primaryKeys
      ? [...new Set([...primaryKeys, primaryKey])]
      : [primaryKey];
    return acc;
  }, {} as Record<string, Primary<T>[]>);
}

export function groupInversedOrMappedKeysByEntity<T extends AnyEntity<T>>(
  collections: readonly Collection<T, AnyEntity>[]
): Record<string, Record<string, Primary<T>[]>> {
  return collections.reduce((acc, cur) => {
    const className = cur.property.type;
    const record = acc[className] || (acc[className] = {});
    // Many to Many vs One to Many
    const inversedOrMappedBy = cur.property.inversedBy
      ? 'inversedBy'
      : cur.property.mappedBy
      ? 'mappedBy'
      : undefined;
    if (!inversedOrMappedBy) {
      throw new Error(
        'Cannot find inversedBy or mappedBy prop: did you forget to set the inverse side of a many-to-many relationship?'
      );
    }
    const key: Primary<T> = getHelperPk(cur.owner.__helper);
    const keys: Primary<T>[] = record[cur.property[inversedOrMappedBy]];
    record[cur.property[inversedOrMappedBy]] = keys
      ? [...new Set([...keys, key])]
      : [key];
    return acc;
  }, {} as Record<string, Record<string, Primary<T>[]>>);
}

export function groupFindQueries<T extends AnyEntity<T>>(
  repoFinds: readonly RepoFind<T>[]
) {
  return repoFinds.reduce((acc, {repo, filter}) => {
    const entityName = repo['entityName'] as string;
    if (acc[entityName]) {
      let matchFound = false;
      const curFilters = acc[entityName].map(el => ({...el}));
      for (const curFilter of curFilters) {
        if (objectsHaveSameKeys(curFilter, filter)) {
          matchFound = true;
          Object.entries(filter).forEach(([prop, idOrIds]) => {
            const curIdOrIds = curFilter[prop];
            if (
              Array.isArray(curIdOrIds) ||
              Array.isArray(idOrIds) ||
              (curIdOrIds && curIdOrIds !== idOrIds)
            ) {
              curFilter[prop] = [
                ...new Set([
                  ...ensureIsArray(curIdOrIds ?? []),
                  ...ensureIsArray(idOrIds),
                ]),
              ];
            } else {
              curFilter[prop] = idOrIds;
            }
          });
          break;
        }
      }
      if (matchFound) {
        acc[entityName] = [...curFilters];
      } else {
        acc[entityName] = [...curFilters, filter];
      }
    } else {
      acc[entityName] = [filter];
    }
    return acc;
  }, {} as Record<string, Record<string, Primary<T> | Primary<T>[]>[]>);
}

export type Result<K> = {
  entityName: string;
  keys: string[];
  entitiesOrError: K[] | Error;
};

export function arrayAreEqual<T extends Primary<unknown>[] | string[]>(
  a: T,
  b: T
): boolean {
  return JSON.stringify(a.sort()) === JSON.stringify(b.sort());
}

export function arrayIncludes<T>(
  arr: Array<Primary<T>>,
  target: Array<Primary<T>>,
  type: 'every' | 'any'
): boolean {
  return !!target[type === 'every' ? 'every' : 'find'](v => arr.includes(v));
}

export function hasRef<T extends AnyEntity<T>>(
  entity: T
): T & Record<string, Reference<T>> {
  return entity as T & Record<string, Reference<T>>;
}

export function hasCol<T extends AnyEntity<T>, K>(
  entity: T
): T & Record<string, Collection<T, K>> {
  return entity as T & Record<string, Collection<T, K>>;
}

export interface IWrappedEntityInternalWithEntity<
  T,
  PK extends keyof T | unknown = PrimaryProperty<T>,
  P = keyof T
> extends IWrappedEntityInternal<T, PK, P> {
  entity: T;
}

export function hasRefOrCol<T extends AnyEntity<T>, K>(
  entity: T
): T & Record<string, Reference<T> | Collection<T, K>> {
  return entity as T & Record<string, Reference<T> | Collection<T, K>>;
}

export function isRef<T extends AnyEntity<T>, K>(
  refOrCol: Reference<T> | Collection<T, K>
): refOrCol is Reference<T> {
  return !(refOrCol instanceof Collection);
}

export function isCol<T extends AnyEntity<T>, K>(
  refOrCol: Reference<T> | Collection<T, K>
): refOrCol is Collection<T, K> {
  return refOrCol instanceof Collection;
}
