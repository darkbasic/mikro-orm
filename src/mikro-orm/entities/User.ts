import {
  Entity,
  Property,
  Enum,
  OneToMany,
  Collection,
  ManyToMany,
} from '@mikro-orm/core';
import {BaseIdEntity, BaseIdEntityConstructor} from './BaseIdEntity';
import {Event} from './Event';

export enum Sex {
  MALE = 0,
  FEMALE = 1,
}

interface UserContructor extends BaseIdEntityConstructor {
  name: string;
  sex: Sex;
}

@Entity()
export class User extends BaseIdEntity {
  @Property()
  name: string;

  @Enum()
  sex: Sex;

  @OneToMany(() => Event, event => event.creator)
  createdEvents: Collection<Event> = new Collection<Event>(this);

  @ManyToMany(() => Event, event => event.partecipants)
  partecipatedEvents: Collection<Event> = new Collection<Event>(this);

  constructor({id, createdAt, name, sex}: UserContructor) {
    super({id, createdAt});
    this.name = name;
    this.sex = sex;
  }
}
