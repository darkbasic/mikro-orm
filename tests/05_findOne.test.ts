import {MikroORM} from '@mikro-orm/core';
import {PostgreSqlDriver} from '@mikro-orm/postgresql';
import {initORMPostgreSql, resetDatabase, sleep} from './utils';
import {User} from '../src/mikro-orm/entities/User';
import {EntityDataLoader} from '../src/dataloader/EntityDataLoader';

describe('FindOne', () => {
  jest.setTimeout(10e3);
  let orm: MikroORM<PostgreSqlDriver>;
  let dataloader: EntityDataLoader;

  beforeAll(async () => {
    orm = await initORMPostgreSql();
    await resetDatabase(orm);
  });

  beforeEach(async () => {
    orm.em.clear();
    dataloader = new EntityDataLoader(orm.em);
    await sleep(100);
  });

  it('should fetch with a single condition', async () => {
    const res = await orm.em.findOne(User, {id: 9});
    expect(res).toBeDefined();
    expect(res).toMatchSnapshot();
  });

  it('should return the same entities with EntityDataLoader', async () => {
    const res = await orm.em.findOne(User, {id: 9});
    const resDataloader = await dataloader.findOne(orm.em.getRepository(User), {
      id: 9,
    });
    expect(res).toBeDefined();
    expect(resDataloader).toBeDefined();
    expect(JSON.stringify(resDataloader)).toBe(JSON.stringify(res));
  });

  const fiveUserIds = [1, 3, 5, 7, 9];

  it('even when queried separately in the same event loop', async () => {
    let res: User[] | string = await Promise.all(
      fiveUserIds.map(id => orm.em.findOneOrFail(User, {id}))
    );
    expect(res).toMatchSnapshot();
    res = JSON.stringify(res);

    orm.em.clear();

    const resDataloader = await Promise.all(
      fiveUserIds.map(id =>
        dataloader.findOneOrFail(orm.em.getRepository(User), {id})
      )
    );
    expect(resDataloader).toBeDefined();
    expect(JSON.stringify(resDataloader)).toBe(res);
  });

  it('should be faster with EntityDataLoader', async () => {
    const start = new Date();
    await Promise.all(fiveUserIds.map(id => orm.em.findOneOrFail(User, {id})));
    const end = new Date();
    const diff = end.getTime() - start.getTime();

    orm.em.clear();

    const startDataloader = new Date();
    await Promise.all(
      fiveUserIds.map(id =>
        dataloader.findOneOrFail(orm.em.getRepository(User), {id})
      )
    );
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThanOrEqual(diff);
  });

  afterEach(async () => {});

  afterAll(async () => {
    orm.close(true);
    await sleep(100);
  });

  afterEach(async () => {});

  afterAll(async () => {
    orm.close(true);
    await sleep(100);
  });
});
