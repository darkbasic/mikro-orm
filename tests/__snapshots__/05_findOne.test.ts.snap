// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindOne even when queried separately in the same event loop 1`] = `
Array [
  Object {
    "createdAt": 2030-02-10T14:45:00.000Z,
    "id": 1,
    "name": "Niccol√≤",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-02-15T08:15:00.000Z,
    "id": 3,
    "name": "Luigi",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-03-02T18:00:00.000Z,
    "id": 5,
    "name": "Riccardo",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-03-06T17:50:00.000Z,
    "id": 7,
    "name": "Sara",
    "sex": 1,
  },
  Object {
    "createdAt": 2030-03-06T21:50:00.000Z,
    "id": 9,
    "name": "Giorgia",
    "sex": 1,
  },
]
`;

exports[`FindOne should fetch with a single condition 1`] = `
Object {
  "createdAt": 2030-03-06T21:50:00.000Z,
  "id": 9,
  "name": "Giorgia",
  "sex": 1,
}
`;
